apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.IntegrationTests/HighPerformancePubsub.cs

```diff
@@ -0,0 +1,163 @@
+            });
+
+            // Publish
+            var rnd = new Random(1234);
+            var activePubs = new HashSet<Task>();
+            int sentCount = 0;
+            long sentSum = 0L; // Sum of bytes of sent messages
+
+            // Watchdog to report progress and fail test on deadlock
+            int prevSentCount = -1;
+            int prevRecvCount = -1;
+            CancellationTokenSource watchdogCts = new CancellationTokenSource();
+            Task.Run(async () =>
+            {
+                while (true)
+                {
+                    await Task.Delay(TimeSpan.FromSeconds(1), watchdogCts.Token).ConfigureAwait(false);
+                    var localSentCount = Interlocked.Add(ref sentCount, 0);
+                    var localRecvCount = Interlocked.Add(ref recvCount, 0);
+                    if (prevSentCount == localSentCount && prevRecvCount == localRecvCount)
+                    {
+                        // Deadlock, shutdown subscriber, and cancel
+                        Console.WriteLine("Deadlock detected. Cancelling test");
+                        bulkSubscriber.StopAsync(new CancellationToken(true));
+                        watchdogCts.Cancel();
+                    }
+                    prevSentCount = localSentCount;
+                    prevRecvCount = localRecvCount;
+                    Console.WriteLine($"Sent: {localSentCount}; Recv: {localRecvCount}");
+                }
+            });
+
+            for (int i = 0; i < messageCount; i++)
+            {
+                if (watchdogCts.IsCancellationRequested)
+                {
+                    Assert.True(false, "Test cancelled by watchdog");
+                }
+                var msgSize = rnd.Next(minMessagesize, maxMessageSize + 1);
+                var msg = new byte[msgSize];
+                rnd.NextBytes(msg);
+                sentSum += msg.Sum(x => (long)x);
+                // Send message, and record Task
+                var pubTask = bulkPublisher.PublishAsync(msg);
+                Interlocked.Increment(ref sentCount);
+                lock (activePubs) activePubs.Add(pubTask);
+                // Remove Task from active when the message has been sent to server
+                Task unused = Task.Run(async () =>
+                {
+                    await pubTask.ConfigureAwait(false);
+                    lock (activePubs) activePubs.Remove(pubTask);
+                });
+                // If too many messages are currently in flight, wait a bit
+                while (activePubs.Locked(() => activePubs.Count) >= maxMessagesInFlight)
+                {
+                    await Task.Delay(TimeSpan.FromMilliseconds(1)).ConfigureAwait(false);
+                }
+            }
+            Console.WriteLine("Publishing complete");
+            // Wait for all messages to be sent to server
+            await Task.WhenAll(activePubs.Locked(() => activePubs.ToArray())).ConfigureAwait(false);
```

Comment:
I've had an instance of an xunit test where it did matter, so I no longer take the chance ;)